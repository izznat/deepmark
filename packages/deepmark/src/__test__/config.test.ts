import np from 'node:path';
import { describe } from 'vitest';
import type { Config, UserConfig } from '../config';
import {
	getSourceDirPaths,
	getSourceFilePaths,
	getOutputDirPaths,
	isFrontmatterFieldIncluded,
	isHtmlElementIncluded,
	isHtmlElementAttributeIncluded,
	isJsonOrYamlPropertyIncluded,
	isJsxComponentIncluded,
	isJsxComponentAttributeIncluded,
	isMarkdownNodeIncluded,
	resolveConfig
} from '../config';

const baseConfig: UserConfig = {
	sourceLanguage: 'en',
	outputLanguages: ['zh'],
	directories: [
		['docs', 'i18n/$langcode$/docs'],
		['blog', 'i18n/$langcode$/blog']
	],
	cwd: np.resolve(process.cwd(), '../../example')
};

describe('resolve paths', (test) => {
	const config: Config = resolveConfig(baseConfig);

	test('resolve source and output directories absolute paths', ({ expect }) => {
		const sourceDirPaths = getSourceDirPaths(config);
		const outputDirPaths = getOutputDirPaths(config);

		expect(sourceDirPaths).toMatchSnapshot();
		expect(outputDirPaths).toMatchSnapshot();
	});

	test('resolve source file absolute paths', async ({ expect }) => {
		const sourceFilePaths = await getSourceFilePaths(config);

		expect(sourceFilePaths).toMatchSnapshot();
	});
});

describe('resolve paths with files include and exclude patterns', (test) => {
	const config: Config = resolveConfig({
		...baseConfig,
		files: {
			include: ['**/blog/**/*'],
			exclude: ['**/blog/*-welcome/**/*']
		}
	});

	test('resolve source file absolute paths', async ({ expect }) => {
		const sourceFilePaths = await getSourceFilePaths(config);

		expect(sourceFilePaths).toMatchSnapshot();
	});
});

describe('default configurations', (test) => {
	const config: Config = resolveConfig(baseConfig);

	test('frontmatter fields', ({ expect }) => {
		expect(isFrontmatterFieldIncluded({ field: 'title', config })).toBe(false);
		expect(isFrontmatterFieldIncluded({ field: 'description', config })).toBe(false);
	});

	test('markdown nodes', ({ expect }) => {
		expect(isMarkdownNodeIncluded({ type: 'code', config })).toBe(false);
		expect(isMarkdownNodeIncluded({ type: 'blockquote', config })).toBe(true);
		expect(isMarkdownNodeIncluded({ type: 'heading', config })).toBe(true);
	});

	test('html elements', ({ expect }) => {
		expect(isHtmlElementIncluded({ tag: 'a', config })).toBe(true);
		expect(isHtmlElementIncluded({ tag: 'div', config })).toBe(true);
	});

	test('html element attributes', ({ expect }) => {
		expect(isHtmlElementAttributeIncluded({ tag: 'div', attribute: 'title', config })).toBe(true);
		expect(isHtmlElementAttributeIncluded({ tag: 'div', attribute: 'id', config })).toBe(false);
	});

	test('jsx components', ({ expect }) => {
		expect(isJsxComponentIncluded({ name: 'Card', config })).toBe(true);
		expect(isJsxComponentIncluded({ name: 'Warning', config })).toBe(true);
	});

	test('jsx component attributes', ({ expect }) => {
		expect(isJsxComponentAttributeIncluded({ name: 'Card', attribute: 'icon', config })).toBe(
			false
		);
		expect(isJsxComponentAttributeIncluded({ name: 'Warning', attribute: 'title', config })).toBe(
			false
		);
	});

	test('json or yaml properties', ({ expect }) => {
		expect(isJsonOrYamlPropertyIncluded({ property: 'author', config })).toBe(false);
		expect(isJsonOrYamlPropertyIncluded({ property: 'title', config })).toBe(false);
	});
});
