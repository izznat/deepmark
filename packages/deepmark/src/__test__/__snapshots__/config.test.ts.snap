// Vitest Snapshot v1

exports[`resolve paths > resolve source and output directories absolute paths 1`] = `
[
  "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$",
  "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs",
  "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog",
]
`;

exports[`resolve paths > resolve source and output directories absolute paths 2`] = `
[
  "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$",
  "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs",
  "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog",
]
`;

exports[`resolve paths > resolve source file absolute paths 1`] = `
{
  "json": [
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/code.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/en/code.json",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docusaurus-plugin-content-blog/options.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/en/docusaurus-plugin-content-blog/options.json",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docusaurus-plugin-content-docs/current.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/en/docusaurus-plugin-content-docs/current.json",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docusaurus-theme-classic/footer.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/en/docusaurus-theme-classic/footer.json",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docusaurus-theme-classic/navbar.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/en/docusaurus-theme-classic/navbar.json",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/_category_.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/_category_.json",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-extras/_category_.json",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-extras/_category_.json",
    },
  ],
  "md": [
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/intro.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/intro.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/congratulations.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/congratulations.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/create-a-blog-post.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/create-a-blog-post.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/create-a-document.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/create-a-document.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/create-a-page.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/create-a-page.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/deploy-your-site.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/deploy-your-site.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-basics/markdown-features.mdx",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-basics/markdown-features.mdx",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-extras/manage-docs-versions.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-extras/manage-docs-versions.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-extras/translate-your-site.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-extras/translate-your-site.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2019-05-28-first-blog-post.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2019-05-28-first-blog-post.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2019-05-29-long-blog-post.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2019-05-29-long-blog-post.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2021-08-01-mdx-blog-post.mdx",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2021-08-01-mdx-blog-post.mdx",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2021-08-26-welcome/index.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2021-08-26-welcome/index.md",
    },
  ],
  "others": [
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-extras/img/docsVersionDropdown.png",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-extras/img/docsVersionDropdown.png",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/docs/tutorial-extras/img/localeDropdown.png",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/docs/tutorial-extras/img/localeDropdown.png",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2021-08-26-welcome/docusaurus-plushie-banner.jpeg",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2021-08-26-welcome/docusaurus-plushie-banner.jpeg",
    },
  ],
  "yaml": [
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/authors.yml",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/authors.yml",
    },
  ],
}
`;

exports[`resolve paths with files include and exclude patterns > resolve source file absolute paths 1`] = `
{
  "json": [],
  "md": [
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2019-05-28-first-blog-post.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2019-05-28-first-blog-post.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2019-05-29-long-blog-post.md",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2019-05-29-long-blog-post.md",
    },
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/2021-08-01-mdx-blog-post.mdx",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/2021-08-01-mdx-blog-post.mdx",
    },
  ],
  "others": [],
  "yaml": [
    {
      "outputFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/i18n/$langcode$/blog/authors.yml",
      "sourceFilePath": "/home/izznatsir/Codes/github/izznatsir/deepmark/example/blog/authors.yml",
    },
  ],
}
`;
